test/Ocean/Hydrostatic_Boussinesq/test_divergence_free.jl

```diff
@@ -0,0 +1,203 @@
+
+let
+  MPI.Initialized() || MPI.Init()
+  mpicomm = MPI.COMM_WORLD
+
+  ll = uppercase(get(ENV, "JULIA_LOG_LEVEL", "INFO"))
+  loglevel = ll == "DEBUG" ? Logging.Debug :
+             ll == "WARN"  ? Logging.Warn  :
+             ll == "ERROR" ? Logging.Error : Logging.Info
+  logger_stream = MPI.Comm_rank(mpicomm) == 0 ? stderr : devnull
+  global_logger(ConsoleLogger(logger_stream, loglevel))
+  @static if haspkg("CUDAnative")
+    device!(MPI.Comm_rank(mpicomm) % length(devices()))
+  end
+
+  @static if haspkg("CuArrays")
+    ArrayType = CuArray
+  else
+    ArrayType = Array
+  end
+
+  brickrange = (xrange, yrange, zrange)
+  topl = StackedBrickTopology(mpicomm, brickrange;
+                              periodicity = (false, false, false),
+                              boundary = ((1, 1), (1, 1), (2, 3)))
+  @show dt = 240 # (Lˣ / cʰ) / Nˣ / N²
+  @show nout = ceil(Int64, tout / dt)
+
+  grid = DiscontinuousSpectralElementGrid(topl,
+                                          FloatType = FT,
+                                          DeviceArray = ArrayType,
+                                          polynomialorder = N,
+                                         )
+
+
+  problem = HSBox{FT}(Lˣ, Lʸ, H, τₒ, fₒ, β)
+
+  model = HBModel{typeof(problem),FT}(problem, cʰ, cʰ, cᶻ, αᵀ, λʳ, νʰ, νᶻ, κʰ, κᶻ)
+
+  dg = OceanDGModel(model,
+                    grid,
+                    Rusanov(),
+                    CentralNumericalFluxDiffusive(),
+                    CentralGradPenalty())
+
+  Q = init_ode_state(dg, FT(0))
+  update_aux!(dg, model, Q, FT(0))
+
+  starttime = Ref(now())
+  cbinfo = GenericCallbacks.EveryXWallTimeSeconds(1, mpicomm) do (s=false)
+    if s
+      starttime[] = now()
+    else
+      energy = norm(Q)
+      maxQ =  Vars{vars_state(model, FT)}(maximum(Q, dims=(1,3)))
+      minQ =  Vars{vars_state(model, FT)}(minimum(Q, dims=(1,3)))
+      @info @sprintf("""Update
+                     simtime = %.16e
+                     runtime = %s
+                     norm(Q) = %.16e
+                     extrema(θ) = (%.16e, %.16e)
+                     Δθ = %.16e
+                     """,
+                     ODESolvers.gettime(lsrk),
+                     Dates.format(convert(Dates.DateTime,
+                                          Dates.now()-starttime[]),
+                                  Dates.dateformat"HH:MM:SS"),
+                     energy,
+                     maxQ.θ,
+                     minQ.θ,
+                     maxQ.θ - minQ.θ)
+      return nothing
+    end
+  end
+
+  lsrk = LSRK144NiegemannDiehlBusch(dg, Q; dt = dt, t0 = 0)
+
+  eng0 = norm(Q)
+  @info @sprintf """Starting
+  norm(Q₀) = %.16e
+  ArrayType = %s""" eng0 ArrayType
+
+  solve!(Q, lsrk, nothing; timeend=timeend, callbacks=(cbinfo,))
+
+  maxQ =  Vars{vars_state(model, FT)}(maximum(Q, dims=(1,3)))
+  minQ =  Vars{vars_state(model, FT)}(minimum(Q, dims=(1,3)))
+
+  @test maxQ.θ ≈ minQ.θ
+  @test maxQ.θ - minQ.θ ≈ 0
```

Comment:
@simonbyrne, @charleskawczynski, I get a "LoadError: scalar getindex is disallowed" error when running on the GPU. Everything worked fine when running on a CPU. Do you guys know why?

Reply:
We may need a `mapreduce` implemented in `MPIStateArrays` for this to work on the GPU

Reply:
Do the current tests run on CPU or GPU?

Reply:
Since you just added the test to [`test/Ocean/runtests.jl`] and not [`slurmci-test.toml`], I believe that this will just be run on the CPU.

I think it would be good to have it running on the GPU too if you can.

[`test/Ocean/runtests.jl`]: https://github.com/climate-machine/CLIMA/blob/4bc0b113570e3e089775b9dad19adb2e529fb80f/test/Ocean/runtests.jln
[`slurmci-test.toml`]: https://github.com/climate-machine/CLIMA/blob/4bc0b113570e3e089775b9dad19adb2e529fb80f/slurmci-test.toml

Reply:
wait, how are the tests decided? Is it not just `test/runtests.jl`? 

Also, I would like this to run on the GPU as well, but am not sure what is needed to make that happen.

Reply:
Azure runs the tests listed in [`test/runtests.jl`](https://github.com/climate-machine/CLIMA/blob/4bc0b113570e3e089775b9dad19adb2e529fb80f/test/runtests.jl) and this is CPU only. Since you added `test/Ocean/runtests.jl` there it will be run on Azure.

If you want test to run on the Caltech cluster (e.g., the runnings tests on the GPU) you need to add them to the [`slurmci-test.toml`] file as well in the relevant section with the correct options.

[`test/runtests.jl`]:https://github.com/climate-machine/CLIMA/blob/4bc0b113570e3e089775b9dad19adb2e529fb80f/test/runtests.jl
[`slurmci-test.toml`]: https://github.com/climate-machine/CLIMA/blob/4bc0b113570e3e089775b9dad19adb2e529fb80f/slurmci-test.toml

Reply:
Okay, I'll keep that in mind for the future.

But still, I know the test will currently fail on the GPU. What exactly do we need to implement to fix that?

Reply:
I think that you'll need to do the initialization on the CPU and someone needs to get `maximum` and `minimum` working with `MPIStateArrays` which are built on `CUArray`

Reply:
I've added support for this in #547.

Reply:
Hi @simonbyrne, I've merged in #547 and #560... but am still getting a map reduce error [1].

Can you take a look at this next week? Thanks for all the help so far

[1] https://www.dropbox.com/s/ypr53qc2pujnngp/stacktrace.txt?dl=0

Reply:
Latest fixes from @simonbyrne work, we're all ready to go.